### PR TITLE
TTT-1889 - Created Expedia model classes for hotel+flight package bat…

### PR DESCRIPTION
**Pull request by LucianTuduce**
…ch search

### CODE REVIEW

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 1]
```diff
@@ -0,0 +1,228 @@
+package uk.co.telegraph.travel.expedia.model
+
+
+case class ExpediaBatchResponse
+(
+  Offers: Seq[Offers],
```
Comment (alin-timu) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Naming on case class parameters should be camel case.

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+  HotelOffers: HotelOffers,
+  FlightOffers: FlightOffers,
+  TransactionId: String
+)
+
+case class Offers
```
Comment (alin-timu) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: This should be called `Offer`

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+  Address: Address,
+  GeoLocation: GeoLocation
+)
+
+case class Occupants
+(
+  Adults: String
+)
+
+case class Offers1
```
Comment (alin-timu) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: I understand why it has `1` in the naming, but I believe that it will be better for the case classes to be split into two files, one for hotel and one for flights. This way the model will be cleaner and you can keep the name `Offer` for the flight model as well.

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 1 above)
Comment (alin-timu) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Discussed about it. It's not worth making them camel case as then it will require a custom serialiser for every parameter.

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 2 above)
Comment (riccardoNovaglia) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Good point

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+
+case class SearchCities
+(
+  Code: String,
+  City: String,
+  Country: String,
+  Province: Option[String]
+)
+
+case class Segments
```
Comment (riccardoNovaglia) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: This class should probably be called `Segment`

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+(
+  Value: String,
+  Currency: String,
+  Name: String
+)
+
+case class FlightOffers
+(
+  Segments: Seq[Segments],
+  Offers: Seq[Offers1],
```
Comment (riccardoNovaglia) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: When we have a list of items, the property is probably items, but the object should probably be Seq[Item]
E.g.
```
...
items: Seq[Item]
...
case class Item(...)
```
or
```
...
items: Items
...
case class Items(items: Seq[Item])
case class Item
```
not
```
... 
items: Seq[Items]
...
case class Items(...)
```
If we have to do specific operations on lists of items it might be sometimes useful to have an aggregate class on top of it, but because these are used to deserialise the messages we receive, it's probably useful to have the first option

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+(
+  Latitude: String,
+  Longitude: String
+)
+
+case class HotelOffers
+(
+  HotelCount: String,
+  NumberOfRooms: String,
+  Occupants: Seq[Occupants],
```
Comment (riccardoNovaglia) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Comment above

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+  Value: String,
+  Currency: String,
+  Name: String
+)
+
+case class FlightOffers
+(
+  Segments: Seq[Segments],
+  Offers: Seq[Offers1],
+  SearchCities: Seq[SearchCities]
```
Comment (riccardoNovaglia) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Comment above

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+case class Fees
+(
+  Value: String,
+  Currency: String,
+  Name: String
+)
+
+case class FlightOffers
+(
+  Segments: Seq[Segments],
```
Comment (riccardoNovaglia) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Comment below

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+  Id: String,
+  Name: String,
+  LocalCurrencyCode: String,
+  Location: Location,
+  Description: Description,
+  ThumbnailUrl: String,
+  StarRating: String,
+  GuestRating: String,
+  GuestReviewCount: String,
+  RoomTypes: Seq[RoomTypes]
```
Comment (riccardoNovaglia) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Comment above

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+  NumberOfNights: String
+)
+
+case class Price
+(
+  PackageBasePrice: PackageBasePrice,
+  PackageTotalPrice: PackageBasePrice,
+  PackageTaxesAndFees: PackageBasePrice,
+  StandAloneTotalPrice: PackageBasePrice,
+  Fees: Seq[Fees],
```
Comment (riccardoNovaglia) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: See comment above

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 11]
```diff
@@ -0,0 +1,228 @@
...
+  FreeWiFi: Boolean,
+  FreeParking: Boolean,
+  FreeBreakfast: Boolean
+)
+
+case class RoomTypes
+(
+  Description: String,
+  RoomKey: String,
+  RatePlans: Seq[RatePlans]
```
Comment (riccardoNovaglia) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: See comment above

File: src/test/scala/uk/co/telegraph/expedia/ExpediaBatchResponseSpec.scala [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+import org.json4s.jackson.JsonMethods
+import org.scalatest.{FreeSpec, Matchers}
+import uk.co.telegraph.dataProviders.ExpediaBatchResponseDataProvider
+import uk.co.telegraph.travel.expedia.model.ExpediaBatchResponse
+
+import scala.io.Source
+
+/**
+  * Tests that the json payload from the file is correctly converted into a {@link ExpediaBatchResponse}.
+  */
```
Comment (riccardoNovaglia) [inline src/test/scala/uk/co/telegraph/expedia/ExpediaBatchResponseSpec.scala]: I don't think we need javadocs for test classes 😄 
The tests themselves should be explicit enough to say what they are and why they exist

File: src/test/scala/uk/co/telegraph/expedia/ExpediaBatchResponseSpec.scala [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+      val expediaBatchResponse = convert(payload)
+
+      expediaBatchResponse should not be null
+      expediaBatchResponse.FlightOffers should not be null
+      expediaBatchResponse.HotelOffers should not be null
+      expediaBatchResponse.Offers.size shouldEqual (2)
+      expediaBatchResponse.HotelOffers.HotelCount shouldEqual "2"
+      expediaBatchResponse.HotelOffers.Hotels.size shouldEqual (2)
+      expediaBatchResponse.FlightOffers.Segments.size shouldEqual (2)
+      expediaBatchResponse.TransactionId shouldEqual ExpediaBatchResponseDataProvider.createExpediaPackages().TransactionId
```
Comment (riccardoNovaglia) [inline src/test/scala/uk/co/telegraph/expedia/ExpediaBatchResponseSpec.scala]: Awesome! I don't remember much of what the content, but consider if there's any particular corner case that would be useful to cover. If there aren't, looks good!

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 2 above)
Comment (LucianTuduce) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Renamed and extracted in separate class

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 8 above)
Comment (LucianTuduce) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Renamed and extracted in separate class

File: src/test/scala/uk/co/telegraph/expedia/ExpediaBatchResponseSpec.scala (hunk 12 above)
Comment (LucianTuduce) [reply src/test/scala/uk/co/telegraph/expedia/ExpediaBatchResponseSpec.scala]: Removed javadoc from test class

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 4 above)
Comment (LucianTuduce) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Renamed and extracted in separate class

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 11 above)
Comment (LucianTuduce) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Renamed and extracted in separate class

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 10 above)
Comment (LucianTuduce) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Renamed and extracted in separate class

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 3 above)
Comment (LucianTuduce) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Renamed and extracted in separate class

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 9 above)
Comment (LucianTuduce) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Renamed and extracted in separate class

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 6 above)
Comment (LucianTuduce) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Renamed and extracted in separate class

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 7 above)
Comment (LucianTuduce) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Renamed and extracted in separate class

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala (hunk 5 above)
Comment (LucianTuduce) [reply src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: Renamed and extracted in separate class

File: src/test/scala/uk/co/telegraph/expedia/ExpediaBatchResponseSpec.scala (hunk 13 above)
Comment (LucianTuduce) [reply src/test/scala/uk/co/telegraph/expedia/ExpediaBatchResponseSpec.scala]: I think this are all the major corner cases.

File: src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+)
+
+
+
+
+
+
+
+
+
```
Comment (riccardoNovaglia) [inline src/main/scala/uk/co/telegraph/travel/expedia/model/ExpediaBatchResponse.scala]: I think we should add more blank lines here 😄

**Comment by LucianTuduce** [comment]
We don't know what properties are needed, as you said, we will filter them later when we know.
